src/ngraph/op/gather.cpp

```diff
@@ -208,3 +210,113 @@ shared_ptr<Node> op::v1::Gather::clone_with_new_inputs(const OutputVector& new_a
+        {
+            out_shape[i] = indices_shape[j];
+        }
+        for (uint64_t j = axis + 1; j < params_shape.size(); i++, j++)
+        {
+            out_shape[i] = params_shape[j];
+        }
+
+        out->set_shape(out_shape);
+
+        if (arg1->get_element_type() == element::i64)
+        {
+            runtime::reference::gather<T, int64_t>(arg0->get_data_ptr<ET>(),
+                                                   arg1->get_data_ptr<int64_t>(),
+                                                   out->get_data_ptr<ET>(),
+                                                   arg0->get_shape(),
+                                                   arg1->get_shape(),
+                                                   out->get_shape(),
+                                                   axis);
+        }
+        else if (arg1->get_element_type() == element::i32)
+        {
+            runtime::reference::gather<T, int32_t>(arg0->get_data_ptr<ET>(),
+                                                   arg1->get_data_ptr<int32_t>(),
+                                                   out->get_data_ptr<ET>(),
+                                                   arg0->get_shape(),
+                                                   arg1->get_shape(),
+                                                   out->get_shape(),
+                                                   axis);
+        }
+        else
+        {
+            throw ngraph_error("Unexpected type");
+        }
+
+        return true;
+    }
+
+    bool evaluate_gather(const HostTensorPtr& arg0,
+                         const HostTensorPtr& arg1,
+                         const HostTensorPtr& out,
+                         size_t axis)
+    {
+        bool rc = true;
+
+        switch (out->get_element_type())
+        {
+            TYPE_CASE(i8)(arg0, arg1, out, axis);
+            break;
+            TYPE_CASE(i16)(arg0, arg1, out, axis);
+            break;
+            TYPE_CASE(i32)(arg0, arg1, out, axis);
+            break;
+            TYPE_CASE(i64)(arg0, arg1, out, axis);
+            break;
+            TYPE_CASE(u8)(arg0, arg1, out, axis);
+            break;
+            TYPE_CASE(u16)(arg0, arg1, out, axis);
+            break;
+            TYPE_CASE(u32)(arg0, arg1, out, axis);
+            break;
+            TYPE_CASE(u64)(arg0, arg1, out, axis);
+            break;
+            TYPE_CASE(f32)(arg0, arg1, out, axis);
+            break;
+            TYPE_CASE(f64)(arg0, arg1, out, axis);
+            break;
+        default: rc = false; break;
+        }
+        return rc;
+    }
+}
+
+bool op::v0::Gather::evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs)
+{
+    return evaluate_gather(inputs[0], inputs[1], outputs[0], get_axis());
+}
+
+bool op::v1::Gather::evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs)
+{
+    int64_t axis = inputs[2]->get_data_ptr<int64_t>()[0];
```

Comment:
The spec doesn't say that the axis input is int64_t and this will do something bad if it's not. Unfortunately, HostTensor doesn't have a `cast_vector` sort of thing. You can do a switch on the element type of inputs[2] and do `*get_data_ptr<element::Type_t::i32>` etc. for each of the integral element types.